lsp/zk.lua

```diff
@@ -1,74 +1,117 @@
 ---@brief
 ---
---- https://github.com/mickael-menu/zk
+--- https://github.com/zk-org/zk
 ---
 --- A plain text note-taking assistant
 
-local function find_zk_root(startpath)
-  for dir in vim.fs.parents(startpath) do
-    if vim.fn.isdirectory(vim.fs.joinpath(dir, '.zk')) == 1 then
-      return dir
+---List notes
+---@param client vim.lsp.Client
+---@param bufnr integer
+---@param opts table
+---@param action fun(path: string, title: string)
+local function zk_list(client, bufnr, opts, action)
+  opts = vim.tbl_extend('keep', { select = { 'path', 'title' } }, opts or {})
+  client:exec_cmd({
+    title = 'ZkList',
+    command = 'zk.list',
+    arguments = { vim.api.nvim_buf_get_name(bufnr), opts },
+  }, { bufnr = bufnr }, function(err, result, _)
+    if err ~= nil then
+      vim.api.nvim_echo({ { 'zk.list error\n' }, { vim.inspect(err) } }, true, {})
+      return
     end
-  end
+    if result == nil then
+      return
+    end
+
+    local paths = vim.tbl_map(function(item)
+      return item.path
+    end, result)
+    local titles = vim.tbl_map(function(item)
+      return item.title
+    end, result)
+    vim.ui.select(titles, {}, function(item, idx)
+      if item ~= nil then
+        action(paths[idx], item)
+      end
+    end)
+  end)
 end
 
 ---@type vim.lsp.Config
 return {
   cmd = { 'zk', 'lsp' },
   filetypes = { 'markdown' },
   root_markers = { '.zk' },
-  ---@param bufnr integer
-  ---@param client vim.lsp.Client
+  workspace_required = true,
   on_attach = function(client, bufnr)
-    vim.api.nvim_buf_create_user_command(bufnr, 'LspZkIndex', function()
+    vim.api.nvim_buf_create_user_command(bufnr, 'LspZkIndex', function(_)
       client:exec_cmd({
         title = 'ZkIndex',
         command = 'zk.index',
         arguments = { vim.api.nvim_buf_get_name(bufnr) },
-      }, { bufnr = bufnr })
-    end, {
-      desc = 'ZkIndex',
-    })
+      }, { bufnr = bufnr }, function(err, result, _)
+        if err ~= nil then
+          vim.api.nvim_echo({ { 'zk.index error\n' }, { vim.inspect(err) } }, true, {})
+          return
+        end
+        if result ~= nil then
+          vim.api.nvim_echo({ { vim.inspect(result) } }, false, {})
+        end
+      end)
+    end, { desc = 'ZkIndex' })
 
-    vim.api.nvim_buf_create_user_command(bufnr, 'LspZkList', function()
-      local bufpath = vim.api.nvim_buf_get_name(bufnr)
-      local root = find_zk_root(bufpath)
+    vim.api.nvim_buf_create_user_command(bufnr, 'LspZkList', function(_)
+      zk_list(client, bufnr, {}, function(path, _)
+        vim.cmd('edit ' .. path)
+      end)
+    end, { desc = 'ZkList' })
 
+    vim.api.nvim_buf_create_user_command(bufnr, 'LspZkTagList', function(_)
```

Comment:
strange to put `_` for a parameter. why not just omit it?

when a parameter is unused, but the one after it is used, then it makes sense to name the unused one something like `_foo`. but i'm not aware of any reason to name a parameter `_` if it is not followed by *used* parameters.